applications/services/gui/modules/number_input.c

```diff
@@ -0,0 +1,389 @@
+                        &I_KeyBackspace_16x9);
+                }
+            } else {
+                if(model->selected_row == row && model->selected_column == column) {
+                    canvas_set_color(canvas, ColorBlack);
+                    canvas_draw_box(
+                        canvas,
+                        keyboard_origin_x + keys[column].x - 3,
+                        keyboard_origin_y + keys[column].y - 10,
+                        11,
+                        13);
+                    canvas_set_color(canvas, ColorWhite);
+                } else if(model->selected_row == -1 && row == 0 && model->selected_column == column) {
+                    canvas_set_color(canvas, ColorBlack);
+                    canvas_draw_frame(
+                        canvas,
+                        keyboard_origin_x + keys[column].x - 3,
+                        keyboard_origin_y + keys[column].y - 10,
+                        11,
+                        13);
+                } else {
+                    canvas_set_color(canvas, ColorBlack);
+                }
+
+                canvas_draw_glyph(
+                    canvas,
+                    keyboard_origin_x + keys[column].x,
+                    keyboard_origin_y + keys[column].y,
+                    keys[column].text);
+            }
+        }
+    }
+}
+
+/** Input callback
+ *
+ * @param      event    The event
+ * @param      context  The context
+ *
+ * @return     true
+ * @return     false
+ */
+static bool number_input_view_input_callback(InputEvent* event, void* context) {
+    NumberInput* number_input = context;
+    furi_assert(number_input);
+
+    bool consumed = false;
+
+    // Fetch the model
+    NumberInputModel* model = view_get_model(number_input->view);
+
+    if(event->type == InputTypeShort || event->type == InputTypeLong ||
+       event->type == InputTypeRepeat) {
+        consumed = true;
+        switch(event->key) {
+        case InputKeyLeft:
+            number_input_handle_left(model);
+            break;
+        case InputKeyRight:
+            number_input_handle_right(model);
+            break;
+        case InputKeyUp:
+            number_input_handle_up(model);
+            break;
+        case InputKeyDown:
+            number_input_handle_down(model);
+            break;
+        case InputKeyOk:
+            number_input_handle_ok(model);
+            break;
+        default:
+            consumed = false;
+            break;
+        }
+    }
+
+    // commit view
+    view_commit_model(number_input->view, consumed);
+
+    return consumed;
+}
+
+void number_input_reset(NumberInput* number_input) {
+    furi_assert(number_input);
+    with_view_model(
+        number_input->view,
+        NumberInputModel * model,
+        {
+            model->header = "";
+            model->selected_row = 0;
+            model->selected_column = 0;
+            model->clear_default_text = false;
+            model->text_buffer = "";
```

Comment:
I got feeling that this part was copied from another place. We are not proud with it )

General concept for view modules is that they should own their own buffer and then provide it in callback call.